sympy/simplify/cse_main.py

```diff
@@ -639,7 +643,7 @@ def _rebuild(expr):
             return expr
 
         if iterable(expr):
-            new_args = [_rebuild(arg) for arg in expr]
```

Comment:
I wonder if `Inverse` is just bugged. If it is considered an iterable, then should this not just work? If you try to iterate `Inverse` then it does nothing (but not give an error either), that seems like weird behaviour to me.

Reply:
Digging into this, the bug appears to occur when rebuilding the `ImmutableDenseMatrix` that is the argument to `Inverse`.
```python
>>> from sympy import ImmutabeDenseMatrix, symbols
>>> A = ImmutableDenseMatrix(symbols('A:4')).reshape(2, 2)
>>> A.args
(3, 3, (A0, A1, A2, A3, A4, A5, A6, A7, A8))
>>> (arg for arg in A)
(A0, A1, A2, A3, A4, A5, A6, A7, A8)
```
Iterating over `A`, as was the previous version of this code results in an invalid constructor being using in `ImmutableDenseMatrix`, which is the source of the error reported in #24689.

Regarding your comment about iterating over `Inverse`: `Inverse` (and other matrix expression types) inherit from `MatExpr` which has the following:
https://github.com/sympy/sympy/blob/26234f080186341318767cfd1c6a602dd43df515/sympy/matrices/expressions/matexpr.py#L57-L60
The fact that nothing is returned, rather than a `TypeError` being raised here is potentially a bug.